File: ios/Sources/Detonator/Element/VerticalScrollViewElement/VerticalScrollViewElement.swift
import UIKit

class VerticalScrollViewElement: Element {
    private var isAtBottom: Bool = false
    
    override public func decodeAttributes(edge: Edge) -> VerticalScrollViewAttributes? {
        return super.decodeAttributes(edge: edge)
    }
    
    override public func createView() -> VerticalScrollView {
        let view = VerticalScrollView()
        
        view.delegate = view
        
        view.isViewGroup = true
        
        view.contentInsetAdjustmentBehavior = .never
        
        view.onPageChangeListener = { page in
            let data = OnPageChangeData(page: page)
            
            self.detonator.emitHandler(name: "onPageChange", edgeId: self.edge.id, data: data)
        }
        
        view.onScrollListener = {
            let child = view.subviews.first!
            
            let diff = child.frame.maxY - (view.frame.size.height + view.contentOffset.y)
            
            self.isAtBottom = diff == 0
        }
        
        return view
    }
    
    override public func patchView() {
        let view = view as! VerticalScrollView
        
        let layoutParams = view.layoutParams
        
        let attributes = attributes as! VerticalScrollViewAttributes
        let prevAttributes = prevAttributes as! VerticalScrollViewAttributes?
        
        let paginated = attributes.paginated
        let prevPaginated = prevAttributes?.paginated
        
        let patchPaginatedBool = forcePatch || paginated != prevPaginated
        
        if patchPaginatedBool {
            view.isPagingEnabled = paginated == true
        }
        
        let inverted = attributes.inverted
        let prevInverted = prevAttributes?.inverted
        
        let patchInvertedBool = forcePatch || inverted != prevInverted
        
        if patchInvertedBool {
            view.inverted = inverted == true
        }
        
        let showsIndicator = attributes.showsIndicator
        let prevShowsIndicator = prevAttributes?.showsIndicator
        
        let patchShowsIndicatorBool = forcePatch || showsIndicator != prevShowsIndicator
        
        if patchShowsIndicatorBool {
            view.showsVerticalScrollIndicator = showsIndicator == true
        }
        
        let scrollToBottom = inverted == true && (forcePatch || isAtBottom);
        
        if scrollToBottom {
            layoutParams.onLayoutClosures["scroll"] = {
                let y = view.contentSize.height - view.frame.size.height
                
                if y > 0 {
                    view.contentOffset = CGPoint(x: 0, y: y)
                }
                
                self.isAtBottom = true
            }
        } else {
            layoutParams.onLayoutClosures["scroll"] = nil
        }
    }
    
    override func patchBorderRadius(
        borderRadius: StyleSize?,
        borderRadiusTopLeft: StyleSize?,
        borderRadiusTopRight: StyleSize?,
        borderRadiusBottomLeft: StyleSize?,
        borderRadiusBottomRight: StyleSize?
    ) {}
    
    struct OnPageChangeData: Encodable {
        let page: Int
    }
    
    class VerticalScrollViewAttributes: Attributes {
        var horizontal: Bool?
        var paginated: Bool?
        var inverted: Bool?
        var showsIndicator: Bool?
        var onPageChange: Bool?
        
        required init(from decoder: Decoder) throws {
            let container = try decoder.container(keyedBy: CodingKeys.self)
            
            horizontal = try container.decodeIfPresent(Bool.self, forKey: .horizontal)
            paginated = try container.decodeIfPresent(Bool.self, forKey: .paginated)
            inverted = try container.decodeIfPresent(Bool.self, forKey: .inverted)
            showsIndicator = try container.decodeIfPresent(Bool.self, forKey: .showsIndicator)
            onPageChange = try container.decodeIfPresent(Bool.self, forKey: .onPageChange)
            
            try super.init(from: decoder)
        }
        
        private enum CodingKeys: String, CodingKey {
            case horizontal
            case paginated
            case inverted
            case showsIndicator
            case onPageChange
        }
    }
}

File: android/Detonator/src/main/java/com/iconshot/detonator/element/verticalscrollviewelement/VerticalScrollViewElement.java
package com.iconshot.detonator.element.verticalscrollviewelement;

import android.view.View;

import com.iconshot.detonator.Detonator;
import com.iconshot.detonator.element.Element;
import com.iconshot.detonator.helpers.CompareHelper;
import com.iconshot.detonator.layout.ViewLayout;

public class VerticalScrollViewElement extends Element<CustomVerticalScrollView, VerticalScrollViewElement.Attributes> {
    private boolean isAtBottom = false;

    public VerticalScrollViewElement(Detonator detonator) {
        super(detonator);
    }

    @Override
    public Class<Attributes> getAttributesClass() {
        return Attributes.class;
    }

    @Override
    public CustomVerticalScrollView createView() {
        CustomVerticalScrollView view = new CustomVerticalScrollView(detonator.context);

        view.setOnPageChangeListener(page -> {
            OnPageChangeData data = new OnPageChangeData();

            data.page = page;

            emitHandler("onPageChange", data);
        });

        view.getViewTreeObserver().addOnScrollChangedListener(() -> {
            View child = view.getChildAt(0);

            int diff = child.getBottom() - (view.getHeight() + view.getScrollY());

            isAtBottom = diff == 0;
        });

        return view;
    }

    @Override
    protected void patchView() {
        ViewLayout.LayoutParams layoutParams = (ViewLayout.LayoutParams) view.getLayoutParams();

        Boolean paginated = attributes.paginated;

        Boolean prevPaginated = prevAttributes != null ? prevAttributes.paginated : null;

        boolean patchPaginated = forcePatch || !CompareHelper.compareObjects(paginated, prevPaginated);

        if (patchPaginated) {
            boolean tmpPaginated = paginated != null ? paginated : false;

            view.setPaginated(tmpPaginated);
        }

        Boolean inverted = attributes.inverted;

        Boolean prevInverted = prevAttributes != null ? prevAttributes.inverted : null;

        boolean patchInverted = forcePatch || !CompareHelper.compareObjects(prevInverted, inverted);

        if (patchInverted) {
            boolean tmpInverted = inverted != null ? inverted : false;

            view.setInverted(tmpInverted);
        }

        Boolean showsIndicator = attributes.showsIndicator;

        Boolean prevShowsIndicator = prevAttributes != null ? prevAttributes.showsIndicator : null;

        boolean patchShowsIndicator = forcePatch || !CompareHelper.compareObjects(showsIndicator, prevShowsIndicator);

        if (patchShowsIndicator) {
            boolean value = showsIndicator != null && showsIndicator;

            view.setVerticalScrollBarEnabled(value);
        }

        boolean scrollToBottom = inverted != null && inverted && (forcePatch || isAtBottom);

        if (scrollToBottom) {
            layoutParams.onLayoutClosures.put("scroll", () -> {
                View child = view.getChildAt(0);

                view.scrollTo(0, child.getBottom());
            });
        } else {
            layoutParams.onLayoutClosures.remove("scroll");
        }
    }

    @Override
    protected void patchPadding(
            Float padding,
            Float paddingHorizontal,
            Float paddingVertical,
            Float paddingTop,
            Float paddingLeft,
            Float paddingBottom,
            Float paddingRight
    ) {
    }

    protected void patchBackgroundColor(
            Object backgroundColor,
            Object borderRadius,
            Object borderRadiusTopLeft,
            Object borderRadiusTopRight,
            Object borderRadiusBottomLeft,
            Object borderRadiusBottomRight
    ) {
        super.patchBackgroundColor(
                backgroundColor,
                borderRadius,
                borderRadiusTopLeft,
                borderRadiusTopRight,
                borderRadiusBottomLeft,
                borderRadiusBottomRight
        );

        ViewLayout.LayoutParams layoutParams = (ViewLayout.LayoutParams) view.getLayoutParams();

        layoutParams.onLayoutClosures.put("borderRadius", () -> {
            float[] radii = getRadii(
                    borderRadius,
                    borderRadiusTopLeft,
                    borderRadiusTopRight,
                    borderRadiusBottomLeft,
                    borderRadiusBottomRight
            );

            view.setRadii(radii);
        });
    }

    private static class OnPageChangeData {
        int page;
    }

    protected static class Attributes extends Element.Attributes {
        Boolean horizontal;
        Boolean paginated;
        Boolean inverted;
        Boolean showsIndicator;
        Boolean onPageChange;
    }
}
